feat: Import cargo-add into cargo

epage opened the pull request:
### Motivation

The reasons I'm aware of are:
- Large interest, see #5586 
- Make it easier to add a dependency when you don't care about the version (instead of having to find it or just using the major version if thats all you remember)
- Provide a guided experience, including
  - Catch or prevent errors earlier in the process
  - Bring the Manifest format documentation into the terminal via `cargo add --help`
  - Using `version` and `path` for `dependencies` but `path` only for `dev-dependencies` (see crate-ci/cargo-release#288 which led to killercup/cargo-edit#480)

### Drawbacks

1. This is another area of consideration for new RFCs, like rust-lang/rfcs#3143 (this PR supports it) or rust-lang/rfcs#2906 (implementing it will require updating `cargo-add`)

2. This is a high UX feature that will draw a lot of attention (ie Issue influx)

e.g.
- killercup/cargo-edit#521
- killercup/cargo-edit#126
- killercup/cargo-edit#217

We've tried to reduce the UX influx by focusing the scope to preserving semantic information (custom sort order, comments, etc) but being opinionated on syntax (style of strings, etc)

### Behavior

Help output
<details>

```console
$ cargo run -- add --help    
cargo-add                                                                                                                                  [4/4594]
Add dependencies to a Cargo.toml manifest file
                                                                         
USAGE:
    cargo add [OPTIONS] <DEP>[@<VERSION>] ...
    cargo add [OPTIONS] --path <PATH> ...
    cargo add [OPTIONS] --git <URL> ...

ARGS:
    <DEP_ID>...    Reference to a package to add as a dependency

OPTIONS:
        --no-default-features     Disable the default features
        --default-features        Re-enable the default features
    -F, --features <FEATURES>     Space-separated list of features to add 
        --optional                Mark the dependency as optional
    -v, --verbose                 Use verbose output (-vv very verbose/build.rs output)
        --no-optional             Mark the dependency as required
        --color <WHEN>            Coloring: auto, always, never
        --rename <NAME>           Rename the dependency
        --frozen                  Require Cargo.lock and cache are up to date
        --manifest-path <PATH>    Path to Cargo.toml
        --locked                  Require Cargo.lock is up to date
    -p, --package <SPEC>          Package to modify
        --offline                 Run without accessing the network
        --config <KEY=VALUE>      Override a configuration value (unstable)
    -q, --quiet                   Do not print cargo log messages
        --dry-run                 Don't actually write the manifest
    -Z <FLAG>                     Unstable (nightly-only) flags to Cargo, see 'cargo -Z help' for
                                  details
    -h, --help                    Print help information

SOURCE:
        --path <PATH>        Filesystem path to local crate to add
        --git <URI>          Git repository location
        --branch <BRANCH>    Git branch to download the crate from
        --tag <TAG>          Git tag to download the crate from
        --rev <REV>          Git reference to download the crate from
        --registry <NAME>    Package registry for this dependency

SECTION:
        --dev                Add as development dependency
        --build              Add as build dependency
        --target <TARGET>    Add as dependency to the given target platform

EXAMPLES:
  $ cargo add regex --build
  $ cargo add trycmd --dev
  $ cargo add --path ./crate/parser/
  $ cargo add serde serde_json -F serde/derive
```

</details>

Example commands
```rust
cargo add regex
cargo add regex serde
cargo add regex@1
cargo add regex@~1.0
cargo add --path ../dependency
```
For an exhaustive set of examples, see [tests](https://github.com/killercup/cargo-edit/blob/merge-add/crates/cargo-add/tests/testsuite/cargo_add.rs) and associated snapshots

Particular points
- Effectively there are two modes
  - Fill in any relevant field for one package
  - Add multiple packages, erroring for fields that are package-specific (`--rename`)
  - Note that `--git` and `--path` only accept multiple packages from that one source
- We infer if the `dependencies` table is sorted and preserve that sorting when adding a new dependency
- Adding a workspace dependency
  - dev-dependencies always use path
  - all other dependencies use version + path
- Behavior is idempotent, allowing you to run `cargo add serde serde_json -F serde/derive` safely if you already had a dependency on `serde` but without `serde_json`
- When a registry dependency's version req is unspecified, we'll first reuse the version req from another dependency section in the manifest.  If that doesn't exist, we'll use the latest version in the registry as the version req

### Additional decisions

Accepting the proposed `cargo-add` as-is assumes the acceptance of the following:
- Add the `-F` short-hand for `--features` to all relevant cargo commands
- Support `@` in pkgids in other commands where we accept `:`
- Add support for `<name>@<version>` in more commands, like `cargo yank` and `cargo install`

### Alternatives

- Use `:` instead of `@` for versions
- Flags like `--features`, `--optional`, `--no-default-features` would be position-sensitive, ie they would only apply to the crate immediate preceding them
  - This removes the dual-mode nature of the command and remove the need for the `+feature` syntax (`cargo add serde -F derive serde_json`)
  - There was concern over the rarity of position-sensitive flags in CLIs for adopting it here
- Support a `--sort` flag to sort the dependencies (existed previously)
  - To keep the scope small, we didn't want general manifest editing capabilities
- `--upgrade <POLICY>` flag to choose constraint (existed previously)
  - The flag was confusing as-is and we feel we should instead encourage people towards `^`
- `--allow-prerelease` so a `cargo add clap` can choose among pre-releases as well
  - We felt the pre-release story is too weak in cargo-generally atm for making it first class in `cargo-add`
- Offer `cargo add serde +derive serde_json` as a shorthand
- Infer path from a positional argument

### Prior Art

- *(Python)* [poetry add](https://python-poetry.org/docs/cli/#add)
  - `git+` is needed for inferring git dependencies, no separate  `--git` flags
  - git branch is specified via a URL fragment, instead of a `--branch`
- *(Javascript)* [yarn add](https://yarnpkg.com/cli/add)
  - `name@data` where data can be version, git (with fragment for branch), etc
  - `-E` / `--exact`, `-T` / `--tilde`, `-C` / `--caret` to control version requirement operator instead of `--upgrade <policy>` (also controlled through `defaultSemverRangePrefix` in config)
  - `--cached` for using the lock file (killercup/cargo-edit#41)
  - In addition to `--dev`, it has `--prefer-dev` which will only add the dependency if it doesn't already exist in `dependencies` as well as `dev-dependencies`
  - `--mode update-lockfile` will ensure the lock file gets updated as well
- *(Javascript)* [pnpm-add](https://pnpm.io/cli/add)
- *(Javascript)* npm doesn't have a native solution
  - Specify version with `@<version>`
  - Also overloads `<name>[@<version>]` with path and repo
    - Supports a git host-specific protocol for shorthand, like `github:user/repo`
    - Uses fragment for git ref, seems to have some kind of special semver syntax for tags?
  - Only supports `--save-exact` / `-E` for operators outside of the default
- *(Go)* [go get](https://go.dev/ref/mod#go-get)
  - Specify version with `@<version>`
  - Remove dependency with `@none`
- *(Haskell)* stack doesn't seem to have a native solution
- *(Julia)* [pkg Add](https://docs.julialang.org/en/v1/stdlib/Pkg/)
- *(Ruby)* [bundle add](https://bundler.io/v2.2/man/bundle-add.1.html)
  - Uses `--version` / `-v` instead of `--vers` (we use `--vers` because of `--version` / `-V`)
  - `--source` instead of `path` (`path` correlates to manifest field)
  - Uses `--git` / `--branch` like `cargo-add`
- *(Dart)* [pub add](https://dart.dev/tools/pub/cmd/pub-add)
  - Uses `--git-url` instead of `--git`
  - Uses `--git-ref` instead of `--branch`, `--tag`, `--rev`

### Future Possibilities

- Update lock file accordingly
- Exploring the idea of a [`--local` flag](https://github.com/killercup/cargo-edit/issues/590)
- Take the MSRV into account when automatically creating version req (https://github.com/killercup/cargo-edit/issues/587)
- Integrate rustsec to report advisories on new dependencies (https://github.com/killercup/cargo-edit/issues/512)
- Integrate with licensing to report license, block add, etc (e.g. https://github.com/killercup/cargo-edit/issues/386)
- Pull version from lock file (https://github.com/killercup/cargo-edit/issues/41)
- Exploring if any vendoring integration would be beneficial (currently errors)
- Upstream `cargo-rm` (#10520), `cargo-upgrade` (#10498), and `cargo-set-version` (in that order of priority)
- Update crates.io with `cargo add` snippets in addition to or replacing the manifest snippets

For more, see https://github.com/killercup/cargo-edit/issues?q=is%3Aissue+is%3Aopen+label%3Acargo-add

### How should we test and review this PR?

This is intentionally broken up into several commits to help reviewing
1. Import of production code from cargo-edit's `merge-add` branch, with only changes made to let it compile (e.g. fixing up of `use` statements).
2. Import of test code / snapshots.  The only changes outside of the import were to add the `snapbox` dev-dependency and to `mod cargo_add` into the testsuite
3. This extends the work in #10425 so I could add back in the color highlighting I had to remove as part of switching `cargo-add` from direct termcolor calls to calling into `Shell`

Structure-wise, this is similar to other commands
- `bin` only defines a CLI and adapts it to an `AddOptions`
- `ops` contains a focused API with everything buried under it

The "op" contains a directory, instead of just a file, because of the amount of content.  Currently, all editing code is contained in there.  Most of this will be broken out and reused when other `cargo-edit` commands are added but holding off on that for now to separate out the editing API discussions from just getting the command in.

Within the github UI, I'd recommend looking at individual commits (and the `merge-add` branch if interested), skipping commit 2.  Commit 2 would be easier to browse locally.

`cargo-add` is mostly covered by end-to-end tests written using `snapbox`, including error cases.  

There is additional cleanup that would ideally happen that was excluded intentionally from this PR to keep it better scoped, including
- Consolidating environment variables for end-to-end tests of `cargo`
- Pulling out the editing API, as previously mentioned
  - Where the editing API should live (`cargo::edit`?)
  - Any more specific naming of types to reduce clashes (e.g. `Dependency` or `Manifest` being fairly generic).
- Possibly sharing `SourceId` creation between `cargo install` and `cargo edit`
- Explore using `snapbox` in more of cargo's tests

Implementation justifications:
- `dunce` and `pathdiff` dependencies: needed for taking paths relative to the user and make them relative to the manifest being edited
- `indexmap` dependency (already a transitive dependency): Useful for preserving uniqueness while preserving order, like with feature values
- `snapbox` dev-dependency: Originally it was used to make it easy to update tests as the UX changed in prep for merging but it had the added benefit of making some UX bugs easier to notice so they got fixed.  Overall, I'd like to see it become the cargo-agnostic version of `cargo-test-support` so there is a larger impact when improvements are made
- `parse_feature` function: `CliFeatures` forces items through a `BTreeSet`, losing the users specified order which we wanted to preserve.

### Additional Information

See also [the internals thread](https://internals.rust-lang.org/t/feedback-on-cargo-add-before-its-merged/16024).

Fixes #5586

rust-highfive commented:
r? @ehuss

(rust-highfive has picked a reviewer for you, use r? to override)

weihanglo commented on src/cargo/ops/cargo_add/dependency.rs:
```diff
@@ -0,0 +1,948 @@
+
+    /// Modify existing entry to match this dependency
+    pub fn update_toml(&self, crate_root: &Path, item: &mut toml_edit::Item) {
+        #[allow(clippy::if_same_then_else)]
```
clippy are not cargo anymore 😆 #10438
```suggestion
        #[allow(clippy::if_same_then_else)]
```

weihanglo reviewed (commented):
Awesome! Really appreciate your great work on integrating cargo-add. I have walked through half of the PR and I think a quick feedback loop is better. Here is a summary of what I've seen so far:

#### Thing I like

- I love how the code is organized! It is self-contained and in a good shape.
- Every public item is documented 👍🏾 
- The coverage of snapshot tests is great!
- Love the write-up of long help messages and the help headings! I can foresee lots of room to improve existing commands 


#### Things I feel uncertain

- IIRC we have a convention that the first line of warning/error messages should be lower-case.
- Some code duplications exist. Such as the logic of selecting a package (cargo-install) or `SourceId` creation. They might be slightly different than the other but look really similar IMO.
- Introduce more ways to get the job done. For example, dunce and snapbox. We need a non-trivial amount of work to migrate from one to another. 

#### Personal preferences

- Use implicit format-arg capture when possible
- I would try to avoid making items/fields public unless they get value to do so.
- I feel like some structs do not need to derive `Clone`.

Great job and thank you @epage!

weihanglo commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,585 @@
+impl<'m> DepTable<'m> {
+    fn to_table(self) -> Vec<&'m str> {
+        match self {
+            Self::Normal => vec!["dependencies"],
+            Self::Development => vec!["dev-dependencies"],
+            Self::Build => vec!["build-dependencies"],
```
Oh. Also `default-features` and `default_features` in a [dependency requirement](https://github.com/rust-lang/cargo/blob/56ccd1f1c79ebfea331a5b7a2effd210e5c9b4db/src/cargo/util/toml/mod.rs#L284-L286)

epage commented on src/bin/cargo/commands/add.rs:
```diff
@@ -0,0 +1,330 @@
+    }
+}
+
+fn parse_section(matches: &ArgMatches) -> DepTable<'_> {
+    if matches.is_present("dev") {
+        DepTable::Development
+    } else if matches.is_present("build") {
+        DepTable::Build
+    } else if let Some(target) = matches.value_of("target") {
+        assert!(!target.is_empty(), "Target specification may not be empty");
+        DepTable::Target(target)
+    } else {
+        DepTable::Normal
+    }
+}
```
We could.

imo the manifest format documentation could use some clean up.  I thought it was odd in cargo-edit and looked up [the docs](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies) but all examples are of `dependencies`, the section is at the same level as all of the different dependency types, and nothing says how they compose together that I saw.

epage commented on src/bin/cargo/commands/add.rs:
```diff
@@ -0,0 +1,330 @@
+            clap::Arg::new("no-default-features")
+                .long("no-default-features")
+                .help("Disable the default features")
+                .long_help(None),
```
This was a carry over from running `cargo expand` on the derive to turn it into the builder.  The derive does it to help in some composition cases.

epage commented on src/bin/cargo/commands/add.rs:
```diff
@@ -0,0 +1,330 @@
+            }
+
+            if let Some(prior) = deps.last_mut() {
+                let features = parse_feature(features);
```
> However, it still makes me feel awkward when I found cargo add tokio +"io sync" is valid. Is it intentional?

imo I'd rather have a single parser everywhere than specialize it based on the application.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,585 @@
+#[derive(Clone, Debug)]
+pub struct AddOptions<'a> {
+    /// Configuration information for cargo operations
+    pub config: &'a Config,
```
1. Several other ops do this
2. While I'm not the biggest SOLID advocate, it would feel dirty / unclear violating single-responsibility by using the workspace to access process-specific configuration.  I can understand accessing that config related to the workspace or for expediency when dealing with deeply nested situations but neither applies here.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,585 @@
+    };
+    let latest = possibilities
+        .iter()
+        .max_by_key(|s| {
+            let stable = s.version().pre.is_empty();
+            (stable, s.version())
+        })
```
If I'm reading this correctly:
- `semver` says `1.0.0 < 2.0.0-alpha.1`
- `cargo-add` says `2.0.0-alpha.1 < 1.0.0`

The goal with this code was to only select a pre-release if that is the only option available.

weihanglo replied:
I am reading the semver crate again. It does state that pre-release versions will be ignored if not specified in `VersionReq` in the doc and its code. Also, each implementation of `Source::query` calls `Dependency::matches_id` and eventually calls `VersionReq::matches`. That being said,

- If prerelease is specified in `VersionReq`, that means the user is fine with picking prerelease version, and semver will pick them if possible.
- If not, semver just ignores them.

Maybe we needn't to check prerelease here and depend on semver for consistency?

- https://github.com/dtolnay/semver/blob/8d8bdb2adaf4572b3c4e8502e7498838b2cf7ccb/src/eval.rs#L3-L24
- https://docs.rs/semver/1.0.6/semver/struct.VersionReq.html#associatedconstant.STAR
- https://github.com/rust-lang/cargo/blob/b816d82ee94acd4d6be588c0f0cd299701c55457/src/cargo/core/source/mod.rs#L32

weihanglo replied:
Maybe we can add a package with version `100000.0.0+my-package` to our tests to prevent regression. The test case [`install::pick_max_version`](https://github.com/rust-lang/cargo/blob/25fcb135d02ea897ce894b67ae021f48107d522b/tests/testsuite/install.rs#L271-L294) already did that.

epage replied:
The important thing to remember is this function's intention is for us to find a version for creating a version-req.  We are in the case where `dep.source().is_none()`.  For example, if you do a `cargo add clap` on an empty crate, we should select 3.1.6 but no pre-release version.  The user can explicitly do so with `cargo add clap@3.0.0-rc.0`.  However, if the crate only exists as pre-release versions, we should allow the user to add that.

epage replied:
> Maybe we can add a package with version 100000.0.0+my-package to our tests to prevent regression. The test case [install::pick_max_version](https://github.com/rust-lang/cargo/blob/25fcb135d02ea897ce894b67ae021f48107d522b/tests/testsuite/install.rs#L271-L294) already did that.

`add_basic` (and many tests using the common package names) currently select from
- 0.1.1
- 0.2.3
- 0.4.1
- 20.0.0
- 999999.0.0 (no idea if I have the number of 9s right)
- 999999.0.0-alpha.1 

Unsure what 100000.0.0+my-package would add that isn't already handled.

I am added a test case that is pre-release only

weihanglo replied:
Ah! I think I got the difference. So basically cargo-add tries harder than cargo-install on finding a usable version. cargo-install avoids prerelease version by [passing a `STAR` VersionReq](https://github.com/rust-lang/cargo/blob/b816d82ee94acd4d6be588c0f0cd299701c55457/src/cargo/ops/cargo_install.rs#L81-L90), whereas cargo-add [gets a `Dependency` with `OptVersionReq::Any` which accepts any versions](https://github.com/rust-lang/cargo/blob/c5f33768a52cc7cffe620ca397ec0fd3b3d2b879/src/cargo/ops/cargo_add/dependency.rs#L213) if unspecified.

Anyway, thanks for your explanation. I am fine with this tiny difference and it indeed makes cargo-add more convenient. 😄

epage commented on src/cargo/ops/cargo_add/dependency.rs:
```diff
@@ -0,0 +1,948 @@
+    }
+
+    /// Get the SourceID for this dependency
+    pub fn source_id(&self, config: &crate::Config) -> CargoResult<crate::core::SourceId> {
```
Also, I'm unsure how well we'll be able to reuse things here.

`cargo-install` directly maps command line arguments to a source-id

`cargo-add` builds up a `Dependency` from multiple sources and then gets the source-id from that.

The role of `Dependency` compared to other items called `Dependency` is its meant to represent the source file for modifying it.  Maybe `ManifestDependency` would clarify that name.

epage replied:
I'm going to assume the +1s were acceptance and I'm marking this as resolved.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,585 @@
+            (stable, s.version())
+        })
+        .ok_or_else(|| {
+            anyhow::format_err!(
+                "The crate `{}` could not be found in registry index.",
+                dependency
+            )
```
Thanks!

epage replied:
Looks like this doesn't actually apply.  This function is called when no version-req is set so we can create a version-req.  It seems rare enough for all possible versions of a crate to be yanked that its not worth adding a special check in here.

epage commented:
@weihanglo 
> Some code duplications exist. Such as the logic of selecting a package (cargo-install) or SourceId creation. They might be slightly different than the other but look really similar IMO.

I'll set aside SourceID as we have a specific conversation on that.

As for selecting a package, I'm going to assume you mean the mapping of arguments to packages
- `cargo-add` supports inline versions with `@` and `cargo-install` doesn't support any inline version as far as I'm aware.  Ideally this gets changed to accepting a pkgid and pkgid gets extended to supporting `@`.  Even at that point, one difference is that `cargo-add` is accepting a version-req while `cargo-install` and `cargo-yank` would accept a version.
- `cargo-add` fills in unspecified arguments from the existing dependency entry.  This makes it harder to reuse.

> Introduce more ways to get the job done. For example, dunce and snapbox. We need a non-trivial amount of work to migrate from one to another.

If there is an alternative to me using `dunce`, I'm willing to change to that and remove the dependency.  Removing it without a replacement though will cause covered test cases to fail.

As for snapbox, I think the value-add is worthwhile based on how its helped identify problems we had missed with previous testing strategies and how its sped up updating tests.  I intended to explore using snapbox in cargo's tests after this and have explicitly listed it in the deferred items.

> Use implicit format-arg capture when possible

Definitely agree for newly written code but almost all of this is written before that was stablized (plus it'll be a hard habit to break until all of my crate's MSRV is new enough...).  If its ok, I'll only focus do this on the side and not make it a focus of my changes.

> I would try to avoid making items/fields public unless they get value to do so.

I generally agree.  Some of the cases were to avoid over-complicating things particularly when dealing with "record" types (data carrying types as compared to abstracted data structures like `Vec`).  In these cases, my preference was to do `pub` all or nothing.

> I feel like some structs do not need to derive Clone.

For the cases of `Clone` in my mind, I disagree with this.  Only implementing fundamental types when they are used, rather than on how they are focusing on how the types are intended to be used, can easily lead to unnecessary code and review churn as they get added/removed based on the current implementation.

epage commented:
All comments I've marked as resolved at this point in time were handled in https://github.com/killercup/cargo-edit/pull/670.  I've just synced those changes back over and force-pushed.

This style of development is an experiment in seeing what works for importing a command.  Normally for a PR I'd do fixups to commits within the PR but since the main commits are meant to mirror specific commits from the cargo-edit repo, I do not want to touch those.  Rather than applying feedback in commits on top of those, I've taken this approach.

epage commented on tests/snapshots/add/target.out/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+name = "cargo-list-test-fixture"
+version = "0.0.0"
+
+[target]
+
+[target.i686-unknown-linux-gnu]
```
Wanted to call out that I have not gotten to this item yet.  I wanted to update this PR with where I was at rather than waiting on any more changes.

Also note that this will only happen when creating the table from scratch.  When the hidden table already exists, we won't change its visibility.

epage commented:
Just pushed a sync of https://github.com/killercup/cargo-edit/pull/671

weihanglo reviewed (commented):
Sorry about flooding lots of comments again. I think we are getting closer!
I'll take a look at most of the test cases soon.

---

BTW I also found a weird case:

```
x = { package = "tokio", version = "1" }
# cargo add rand --rename x --optional
x = { package = "tokio", version = "1", optional = true }
```

It looks confusing to me but I am not sure what is the expected behaviour?

epage commented on src/cargo/ops/cargo_add/dependency.rs:
```diff
@@ -0,0 +1,987 @@
+                    } else {
+                        table.remove("rev");
+                    }
+                    for key in ["version", "path"] {
+                        table.remove(key);
+                    }
```
This is one of the listed reasons why `git` support is feature gated.  We didn't want to block cargo-add support on finishing it up.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,553 @@
+        }
+
+        print_msg(&mut options.config.shell(), &dep, &dep_table)?;
+        if let Some(Source::Path(src)) = dep.source() {
+            if src.path == manifest.path.parent().unwrap_or_else(|| Path::new("")) {
+                anyhow::bail!(
+                    "cannot add `{}` as a dependency to itself",
+                    manifest.package_name()?
+                )
+            }
+        }
```
Tried it out but instead I'm going to make the order
- print_msg
- check error
- feature warning

`print_msg` says `Adding cargo-list-test-fixture-dependency (local) to dev-dependencies.` and it feels less abrupt / friendlier to see that message before seeing an error about about (might be different if it was "Added" instead of "Adding")

epage commented on src/bin/cargo/commands/add.rs:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    if crates.len() > 1 && rename.is_some() {
+        anyhow::bail!("cannot specify multiple crates with rename");
```
Good catch!  Forgot to adjust this when we added inline features

epage commented on src/cargo/ops/cargo_add/crate_spec.rs:
```diff
@@ -0,0 +1,87 @@
+            Self::Path(path) => {
+                let manifest = get_manifest_from_path(path)?;
+                let crate_name = manifest.package_name()?;
+                let version = manifest.package_version()?;
+                let path = dunce::canonicalize(path)?;
+                let available_features = manifest.features()?;
+                Dependency::new(crate_name)
+                    .set_source(PathSource::new(path).set_version(version))
```
We could preserve the lack of version but it'll be some extra complication that, at the moment, I'm unsure if its worth.  When we merge the existing dep with the CLI dep, we preserve the CLI dep's source to make sure we get the version the user specified.  We'd need to specialize this based on CLI args.
```rust
    let mut dependency = if let Some(mut old_dep) = old_dep.clone() {
        if spec_dep.source().is_some() {
            // Overwrite with `crate_spec`
            old_dep.source = spec_dep.source;
        }
        old_dep = populate_dependency(old_dep, arg);
        old_dep
    } else {
        spec_dep
    };
```

We do automatically strip versions when using a dev-dependency.   I've considered doing the same for crates with `publish = false`.

weihanglo replied:
Actually, this is also an edge case and there is no harm in adding the version back.
(unless it is used to prevent accidental publishing but users should definitely use `publish = false` instead 😂)

Well, I'm sold. Let's skip it for now.

epage replied:
Just so this gets recorded, benefits for not including `version`:
- In dev-dependencies, skipping `version`  makes it easier for tools like `cargo-release` to publish cyclic dependencies
- In any dependency, skipping `version` reduces changes needed when bumping versions, automatic or manual (e.g. I have a workspace with some "cargo xtasks" that aren't published mixed with crates that are but all dependency versions have to be bumped).

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,553 @@
+    let legacy = manifest.get_legacy_sections();
+    if !legacy.is_empty() {
+        anyhow::bail!(
+            "Deprecated dependency sections are unsupported: {}",
```
Maybe I don't have a good sense of how often these are used in the wild but I'm thinking its not worth the effort of mapping the legacy key to the modern version and rendering them to be comprehendable.

epage commented:
I just pushed https://github.com/killercup/cargo-edit/pull/672 which has everything that I marked as resolved in the most recent batch of feedback.

weihanglo commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,560 @@
+    let mut dependency = if let Some(mut old_dep) = old_dep.clone() {
+        if old_dep.name != spec_dep.name {
+            // Assuming most existing keys are not relevant when the package changes
+            if spec_dep.optional.is_none() {
+                spec_dep.optional = old_dep.optional;
+            }
```
Why only `optional` is preserved?

epage replied:
This is when we switch
```toml
some-package = { package = "my-package1", version = "0.1.1", optional = true }
```
to
```toml
some-package = { package = "my-package2", version = "99999.0.0", optional = true }
```

My assumption is that the underlying crate we re-targeted to might have different features, version, registry, etc.  The only thing that remains the same is how its used within the current crate, just the dependency key and `optional`.

weihanglo replied:
Actually, I am not sure about the whole thing of overwriting an existing renamed dep with the other package. There is no clue what's going on in stderr. Do you think it is worth a warning or something else?

weihanglo reviewed (commented):
Two snapshots seem not used anymore:

- cargo_config_source_empty
- git_external

weihanglo commented:
So far the integration looks good to me. Highlight some good parts here again:

- The whole `cargo-add` command is self-contained. There is nearly no interaction with other existing parts. it's very pleasant to walk through the PR even with six thousand line changes.
- Snapshot tests with [`snapbox`]: This really improves the overall experience for both developers and reviewers. There are still [some missing features](zulip-snapbox-diff) in snapbox that we should figure out, but I don't think it's a blocker. 
- The development flow of importing an external command, [which @epage mentioned here], works well for me. This can be a good reference for future integrations.

Besides what has been mentioned in the PR description, there is an item that needs to tackle in the follow-up:

- [ ] Manpage of `cargo-add` (triggered via `cargo help add`)

And two unrelated enhancements for cargo internals:

- The road to `snapbox`: Maybe we can have a guide to porting tests and call out for help?
- Continue fixing messy path issues rooted in cargo with `dunce` and `pathdiff`?

Anyway, I guess it's time to see if we can reach a consensus.

@rfcbot merge


[`snapbox`]: https://crates.io/crates/snapbox
[zulip-snapbox-diff]: https://zulip-archive.rust-lang.org/stream/246057-t-cargo/topic/snapbox.20diff.html
[which @epage mentioned here]: https://github.com/rust-lang/cargo/pull/10472#issuecomment-1067506228

weihanglo commented:
So far the integration looks good to me. Highlight some good parts here again:

- The whole `cargo-add` command is self-contained. There is nearly no interaction with other existing parts. it's very pleasant to walk through the PR even with six thousand line changes.
- Snapshot tests with [`snapbox`]: This really improves the overall experience for both developers and reviewers. There are still [some missing features](zulip-snapbox-diff) in snapbox that we should figure out, but I don't think it's a blocker. 
- The development flow of importing an external command, [which @epage mentioned here], works well for me. This can be a good reference for future integrations.

Besides what has been mentioned in the PR description, there is an item that needs to tackle in the follow-up:

- [ ] Manpage of `cargo-add` (triggered via `cargo help add`)

And two unrelated enhancements for cargo internals:

- The road to `snapbox`: Maybe we can have a guide to porting tests and call out for help?
- Continue fixing messy path issues rooted in cargo with `dunce` and `pathdiff`?

Anyway, I guess it's time to see if we can reach a consensus.

@rfcbot fcp merge


[`snapbox`]: https://crates.io/crates/snapbox
[zulip-snapbox-diff]: https://zulip-archive.rust-lang.org/stream/246057-t-cargo/topic/snapbox.20diff.html
[which @epage mentioned here]: https://github.com/rust-lang/cargo/pull/10472#issuecomment-1067506228

weihanglo commented:
@rfcbot fcp merge

weihanglo commented:
@rfcbot fcp merge

epage commented:
Just wanted to highlight that approving `cargo add` is also approving:
- Add the `-F` short-hand for `--features` to all relevant cargo commands
- Support `@` in pkgids in other commands where we accept `:`
- Add support for `<name>@<version>` in more commands, like `cargo yank` and `cargo install`

epage commented:
As an aside
> The whole cargo-add command is self-contained. There is nearly no interaction with other existing parts. it's very pleasant to walk through the PR even with six thousand lines of addition.
> ...
> The development flow of importing an external command, https://github.com/rust-lang/cargo/pull/10472#issuecomment-1067506228, works well for me. This can be a good reference for future integrations.

I wonder if another way of modularizing cargo is to take a "porcelain" and "plumbing" model and have the `cargo` binary focus on "plumbing" where possible and have most commands external, like we have `cargo fmt` and `cargo clippy` as external today.

That'd also get us to dog food the command-line API.

Eh2406 commented:
@rfcbot concern insta-stable

We don't want to make this unstable and make it hard for people to use their existing `cargo add`.
But this is a very big change to get no unstable time.
Discussion happening at: https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Stablizing.20cargo-add

epage commented:
Updated the code and I think I've updated everything in the PR description
- `cargo add serde serde_json -F serde/derive` instead of `cargo add serde +derive serde_json`
- `cargo add --git <url>` is now supported.  Crate name is optional if there is only one in the repo.  Multiple crate names within the repo are supported.
- `cargo add --path <path>` instead of `cargo add <path>`.  Cannot be mixed with adding from the registry or git
- Removed some short flags to be more conservative in the first release
- Now supports git+registry dependencies, like path

ehuss reviewed (commented):
(Just a partial review.)

I'm a bit concerned about using symlinks in the test directory.  On Windows, it requires developer mode and `core.symlinks` enabled.  I pinged on #t-infra on Zulip to see if anyone had any thoughts.  This might cause some problems for people wanting to work on cargo.  I think at a minimum there should be some documentation at https://doc.crates.io/contrib/process/working-on-cargo.html#building-cargo that points to the microsoft docs on enabling developer mode and checking core.symlinks.

This seems to be missing documentation.  There are a few places to add the documentation:
* Add the command here: https://github.com/rust-lang/cargo/tree/master/src/doc/src/commands
* Add to [SUMMARY.md](https://github.com/rust-lang/cargo/blob/master/src/doc/src/SUMMARY.md)
* Add to list in [manifest commands](https://github.com/rust-lang/cargo/blob/master/src/doc/src/commands/manifest-commands.md)
* [Rebuild the man pages](https://github.com/rust-lang/cargo/tree/master/src/doc#building-the-man-pages)

There's also the completion files:
* zsh: https://github.com/rust-lang/cargo/blob/master/src/etc/_cargo
* bash: https://github.com/rust-lang/cargo/blob/master/src/etc/cargo.bashcomp.sh

epage commented:
> I'm a bit concerned about using symlinks in the test directory. On Windows, it requires developer mode and core.symlinks enabled. I pinged on #t-infra on Zulip to see if anyone had any thoughts. This might cause some problems for people wanting to work on cargo. I think at a minimum there should be some documentation at https://doc.crates.io/contrib/process/working-on-cargo.html#building-cargo that points to the microsoft docs on enabling developer mode and checking core.symlinks.

This has been tested on Windows without symlink support; `snapbox` will read the files git puts in place of symlinks.

> This seems to be missing documentation. There are a few places to add the documentation:

So I was not aware of all of these (should have looked closer at cargo-tree) but I was at least calling out man pages as something deferred out of this PR.  It'd greatly reduce churn if we have agreed on what the CLI should look like before hand writing all of that documentation / completions.

btw is there a reason we don't have clap generate completions?  We can set it up to generate completions either at runtime or at development time (with a test ensuring the completions aren't stale)

weihanglo commented:
Oh. Just found once case is not covered by test. Could we add one?

```
$ cargo add --path ./dep --features dep/feat
error: `dep/feat` is unsupported when inferring the crate name, use `feat`
```

epage commented on src/cargo/ops/cargo_add/dependency.rs:
```diff
@@ -0,0 +1,1026 @@
+                        table.remove("version");
+                    }
+
+                    for key in ["path"] {
+                        table.remove(key);
+                    }
```
I had considered this but I preferred the symmetry between the different sources, making what is happening more obvious.

epage commented:
I've updated the PR, see https://github.com/killercup/cargo-edit/pull/689 for the more incremental changes.

ehuss commented:
> btw is there a reason we don't have clap generate completions? We can set it up to generate completions either at runtime or at development time (with a test ensuring the completions aren't stale)

I haven't looked at clap completions for a while, but the completions have dynamic functions for looking up values. Does it support that?

The plan (#6645) was to switch completions to be driven by cargo itself (similar in style to how npm handles completions).  I just never got around to it.

epage commented:
> > btw is there a reason we don't have clap generate completions? We can set it up to generate completions either at runtime or at development time (with a test ensuring the completions aren't stale)
> 
> I haven't looked at clap completions for a while, but the completions have dynamic functions for looking up values. Does it support that?
> 
> The plan (#6645) was to switch completions to be driven by cargo itself (similar in style to how npm handles completions). I just never got around to it.

https://github.com/clap-rs/clap/issues/1232 is the relevant issue for clap.  We plan to have completions determined in Rust code and then a layer will report it for the relevant shell.  We'll then allow people to provide custom completers for arguments.

As a way of reducing cargo maintenance burden, maybe I'll make this my next priority within clap.

weihanglo reviewed (approved):
The latest change looks good to me

bors commented:
:umbrella: The latest upstream changes (presumably #10538) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ehuss commented on tests/snapshots/copy.sh:
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
```
It's not really clear what these shell scripts are being added for.  Are these things people would be expected to use?  Can they include some comments at least providing some information on what they are for?

epage replied:
Sure, I'll had comments.

ehuss commented on tests/testsuite/cargo_add.rs:
```diff
@@ -0,0 +1,2114 @@
+    snapbox::cmd::cargo_bin!("cargo")
+}
+
+pub fn cargo_command() -> snapbox::cmd::Command {
```
This seems really unfortunate to duplicate all the code from cargo-test-support.  Can this be shared somehow?

epage replied:
Yes, I did this for having a cleaner integration of the code for this PR and have it in the deferred section for cleaning up in a follow up PR.

ehuss reviewed (commented):
I am finding it hard to read and understand the tests when the code for running the test is separate from the setup and expected output.  I don't suppose there is any way to make that better?  For example, in compiletest there are [headers](https://rustc-dev-guide.rust-lang.org/tests/headers.html) which are used to drive the test.  That might be very difficult here since cargo tests tend to be quite a bit more complicated, so that may not work well enough.  Overall I'm finding it quite difficult to review, though that may be a factor of the sheer number of tests.  Anyway, not blocking, just leaving some feedback.

Can you add a chapter to the contributor guide similar to https://doc.crates.io/contrib/tests/writing.html that explains how to write and use the snapshot tests?  What kind of automated support is there for writing and updating them?

This looks good to me barring the documentation.

epage commented:
> I am finding it hard to read and understand the tests when the code for running the test is separate from the setup and expected output. I don't suppose there is any way to make that better? For example, in compiletest there are [headers](https://rustc-dev-guide.rust-lang.org/tests/headers.html) which are used to drive the test. That might be very difficult here since cargo tests tend to be quite a bit more complicated, so that may not work well enough. Overall I'm finding it quite difficult to review, though that may be a factor of the sheer number of tests. Anyway, not blocking, just leaving some feedback.

When using `trycmd`, you have a TOML file with the command, stdin, stdout, and stderr that sits next to the directories.  I had to drop down a layer in the testing stack to snapbox because we needed more dynamic tests, like git directories being created and the tests referencing them.

Also, one of the aims is for the snapshots to be as pristine as possible so you can still operate on their files, which is why I've avoided headers like insta.

> Can you add a chapter to the contributor guide similar to https://doc.crates.io/contrib/tests/writing.html that explains how to write and use the snapshot tests? What kind of automated support is there for writing and updating them?

Right now there are several limitations for using snapbox inside of cargo.  Should we wait until some of those restrictions are lifted?  It might be good to port some tests over first to help verify what its like.

> This looks good to me barring the documentation.

Is this referring to the new conversations started or the past one about man and others.  I had asked that we do those in a follow up PR but haven't seen a comment one way or the other on that idea.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -0,0 +1,650 @@
+                options
+                    .config
+                    .shell()
+                    .warn(format!("unrecognized features: {unknown_features:?}"))?;
```
It started off as a warning as I didn't have full confidence in the feature detection matching cargo.  With some bug reports and integrating with cargo, my confidence has gone up, so we could switch this to an error now and the chance of someone being blocked is relatively low and they can work around it.

ehuss replied:
Yea, I think the implementation is fine, and I would feel a little more comfortable with an error.

ehuss commented:
> Should we wait until some of those restrictions are lifted?

I don't know what those restrictions are.  Are the tests generated manually?  One way or the other, I would expect some documentation on how to write these tests, even if it is just basic instructions.

> Is this referring to the new conversations started or the past one about man and others. I had asked that we do those in a follow up PR but haven't seen a comment one way or the other on that idea.

Ah, sorry, I missed that.  I think it would be fine to do that in a separate PR as long as it is done relatively soon.

epage commented:
https://github.com/killercup/cargo-edit/pull/692 represents updates from the latest review feedback

> I don't know what those restrictions are. Are the tests generated manually? One way or the other, I would expect some documentation on how to write these tests, even if it is just basic instructions.

1. Until I've done the PR that generalizes this, the code for using snapshots with cargo is in the cargo_add tests and it'd need to be duplicated into other tests
2. This only covers matching-with-substitutions and doesn't cover the unordered or json cases.  Either someone starts down one testing method and has to rewrite when they come across something they need, we'd need to write up how to use the existing low-level cargo-test-support code, or we wait.
3. Having just looked at the docs, and I feel like we'd need to bifurcate them between the two approaches.  Personally, I'd just shift the focus when the new approach is "ready enough".

> Ah, sorry, I missed that. I think it would be fine to do that in a separate PR as long as it is done relatively soon.

That is my intent.  The follow ups will be my top priority.

Eh2406 commented:
@rfcbot resolve insta-stable

ehuss commented:
🎉 🎉 🎉 🎉 🎉 

I discussed with @epage about following up on documentation and other things that were deferred here.

Thanks so much @epage for putting in the massive work for this, and thanks @weihanglo for all the reviews!

@bors r+

bors commented:
:hourglass: Testing commit a970bfc11c2483ab1b3663b7f9430d30015d7ed7 with merge 6a4d98d2327124ca52bb9f67d6ad0097eb6b2e65...
<!-- homu: {"type":"BuildStarted","head_sha":"a970bfc11c2483ab1b3663b7f9430d30015d7ed7","merge_sha":"6a4d98d2327124ca52bb9f67d6ad0097eb6b2e65"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/6067135633?check_suite_focus=true)
Approved by: ehuss
Pushing 6a4d98d2327124ca52bb9f67d6ad0097eb6b2e65 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/6067135633?check_suite_focus=true"},"merge_sha":"6a4d98d2327124ca52bb9f67d6ad0097eb6b2e65"} -->

stonecharioteer commented:
I am a silent lurker in these issues, being new to rust but God this is a beautiful PR all. Just so well documented and discussed. There's so much to learn. Apologies for being off topic.

weihanglo commented:
> > I'm a bit concerned about using symlinks in the test directory. On Windows, it requires developer mode and core.symlinks enabled. I pinged on #t-infra on Zulip to see if anyone had any thoughts. This might cause some problems for people wanting to work on cargo. I think at a minimum there should be some documentation at https://doc.crates.io/contrib/process/working-on-cargo.html#building-cargo that points to the microsoft docs on enabling developer mode and checking core.symlinks.
> 
> This has been tested on Windows without symlink support; `snapbox` will read the files git puts in place of symlinks.

@epage, does this still hold today? I hit this error today on windows-msvc machine. By setting `core.symlinks = true` globally and re-cloning the repo the issue was fixed.

weihanglo commented:
For the symlink error, here are steps to reproduce:

* Create a Windows Server 2022 on AWS EC2 (also failed on my own Windows 11 developer preview)
* Install C++ build tools
* Install Git for Windows (with symlink feature disabled)
* clone rust-lang/cargo
* `cargo t --test testsuite cargo_add::add_basic`

Feel free to copy this to snapbox issue. Thank you!

```console
running 1 test
test cargo_add::add_basic::case ... FAILED

failures:


---- cargo_add::add_basic::case stdout ----
thread 'cargo_add::add_basic::case' panicked at 'called `Result::unwrap()` on an `Err` value: Error { inner: "Failed to copy C:\\Users\\Administrator\\Documents\\cargo\\tests\\testsuite\\cargo_add\\add_basic\\in to C:\\Users\\Administrator\\Documents\\cargo\\target\\tmp\\cit\\t0\\case\\: The system cannot find the path specified. (os error 3)", backtrace: None }', C:\Users\Administrator\Documents\cargo\crates\cargo-test-support\src\lib.rs:324:77
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

```console
PS C:\Users\Administrator\Documents\cargo> ls .\tests\testsuite\cargo_add\add_basic\

    Directory: C:\Users\Administrator\Documents\cargo\tests\testsuite\cargo_add\add_basic

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----          3/8/2023   8:18 PM                out
-a----          3/8/2023   8:18 PM             15 in
-a----          3/8/2023   8:18 PM            744 mod.rs
-a----          3/8/2023   8:18 PM             90 stderr.log
-a----          3/8/2023   8:18 PM              0 stdout.log

PS C:\Users\Administrator\Documents\cargo> cat .\tests\testsuite\cargo_add\add_basic\in
../add-basic.in
```

```
cargo 1.67.1 (8ecd4f20a 2023-01-10)
release: 1.67.1
commit-hash: 8ecd4f20a9efb626975ac18a016d480dc7183d9b
commit-date: 2023-01-10
host: x86_64-pc-windows-msvc
libgit2: 1.5.0 (sys:0.16.0 vendored)
libcurl: 7.86.0-DEV (sys:0.4.59+curl-7.86.0 vendored ssl:Schannel)
os: Windows 10.0.20348 (Windows Server 2022 Datacenter) [64-bit]
```
